src/protocols/bdx/BDXMessageUtils.cpp

```diff
@@ -0,0 +1,408 @@
+    uint8_t ptcByte;
+    uint8_t rangeCtl;
+    bool hasStartOffset     = false;
+    uint32_t tmpUint32Value = 0;
+    uint8_t * bufStart      = aBuffer.Start();
+    Reader bufReader(bufStart, aBuffer.DataLength());
+
+    SuccessOrExit(bufReader.Read8(&ptcByte).Read8(&rangeCtl).Read16(&aParsedMessage.mMaxBlockSize).StatusCode());
+
+    aParsedMessage.mSupportedVersions     = ptcByte & VERSION_MASK;
+    aParsedMessage.mSupportsSenderDrive   = ((ptcByte & SENDER_DRIVE_MASK) != 0);
+    aParsedMessage.mSupportsReceiverDrive = ((ptcByte & RECEIVER_DRIVE_MASK) != 0);
+    aParsedMessage.mSupportsAsync         = ((ptcByte & ASYNC_MASK) != 0);
+
+    aParsedMessage.mDefLen    = (rangeCtl & DEFLEN_MASK) != 0;
+    hasStartOffset            = (rangeCtl & START_OFFSET_MASK) != 0;
+    aParsedMessage.mWideRange = (rangeCtl & WIDERANGE_MASK) != 0;
+
+    if (hasStartOffset)
+    {
+        if (aParsedMessage.mWideRange)
+        {
+            SuccessOrExit(bufReader.Read64(&aParsedMessage.mStartOffset).StatusCode());
+        }
+        else
+        {
+            SuccessOrExit(bufReader.Read32(&tmpUint32Value).StatusCode());
+            aParsedMessage.mStartOffset = tmpUint32Value;
+        }
+    }
+
+    if (aParsedMessage.mDefLen)
+    {
+        if (aParsedMessage.mWideRange)
+        {
+            SuccessOrExit(bufReader.Read64(&aParsedMessage.mMaxLength).StatusCode());
+        }
+        else
+        {
+            SuccessOrExit(bufReader.Read32(&tmpUint32Value).StatusCode());
+            aParsedMessage.mMaxLength = tmpUint32Value;
+        }
+    }
+
+    SuccessOrExit(bufReader.Read16(&aParsedMessage.mFileDesLength).StatusCode());
+
+    VerifyOrExit(bufReader.OctetsRead() + aParsedMessage.mFileDesLength <= aBuffer.DataLength(),
+                 err = CHIP_ERROR_MESSAGE_INCOMPLETE);
+    aParsedMessage.mFileDesignator = &bufStart[bufReader.OctetsRead()];
+
+    // Rest of message is metadata (could be empty)
+    if (bufReader.OctetsRead() + aParsedMessage.mFileDesLength < aBuffer.DataLength())
+    {
+        uint16_t metadataStartIndex    = static_cast<uint16_t>(bufReader.OctetsRead() + aParsedMessage.mFileDesLength);
+        aParsedMessage.mMetadata       = &bufStart[metadataStartIndex];
+        aParsedMessage.mMetadataLength = static_cast<uint16_t>(aBuffer.DataLength() - metadataStartIndex);
+    }
+
+exit:
+    if (bufReader.StatusCode() != CHIP_NO_ERROR)
+    {
+        err = bufReader.StatusCode();
+    }
+    return err;
+}
+
+size_t TransferInit::PackedSize()
+{
+    size_t offsetSize = mWideRange ? 8 : 4;
+    size_t lengthSize = offsetSize;
+
+    // First 2 bytes are Transfer Control and Range Control bytes
+    return (2 + sizeof(mMaxBlockSize) + offsetSize + lengthSize + sizeof(mFileDesLength) + mFileDesLength + mMetadataLength);
+}
+
+bool TransferInit::operator==(const TransferInit & another) const
+{
+    bool fileDesMatches  = memcmp(mFileDesignator, another.mFileDesignator, mFileDesLength) == 0;
```

Comment:
need to check that lengths are the same.
This is before you do memcmp since maybe another.mFileDesignator memory is not so long as to allow you to compare (most mcus will probably not care, but boy would this be hard to track down)

Reply:
Will do